### PR TITLE
[WIP] Finalizing/Testing of TwistController features

### PR DESCRIPTION
**Pull request by fmessmer**
This is a WIP/CodeReview PR that still needs some more tweaking. Do not merge yet!
(Mainly motivated by @ipa-fxm-mb finalizing his thesis soon)

Also, all remaining issues regarding package_xml_migration will be handled here for the following packages (as I keep merging back and forth all the time): 
- cob_cartesian_controller
- cob_frame_tracker
- cob_obstacle_distance
- cob_twist_controller

@ipa-fxm-mb FYI

### CODE REVIEW

**Comment by fmessmer** [comment]
Things to fix for package_xml_migration:
- explicit declaration of dependency to `boost` (including proper COMPONENTS)
- review install-tags for headers/libraries if not needed
- cleanup `catkin_package` exports accordingly

**Comment by fmessmer** [comment]
add install tags for libraries

File: cob_cartesian_controller/scripts/action_test.py [hunk 1]
```diff
@@ -0,0 +1,425 @@
...
+#    goal.move_lin.roll = 0
+#    goal.move_lin.pitch = 0
+#    goal.move_lin.yaw = 0
+#    goal.move_lin.rotate_only = False
+#
+#    goal.move_lin.profile.profile.vel = 0.2
+#    goal.move_lin.profile.profile.accl = 0.1
+#    goal.move_lin.profile.profile_type = Profile.SINOID
+
+if __name__ == '__main__':
```
Comment (fmessmer) [inline cob_cartesian_controller/scripts/action_test.py]: Do we still use this script? Or is it superseded by those in `cob_twist_controller/scripts/test`?
If it's not needed, I'd remove it...or at least split it maybe...

**Comment by fmessmer** [comment]
For enhancements/open issues regarding `cob_cartesian_controller`, I created a new issue #52 
They will be addressed in a new PR in order not to hinder this one from being merged.

Same for enhancements of `cob_twist_controller`, New issue https://github.com/ipa320/cob_control/issues/53

File: cob_frame_tracker/src/cob_frame_tracker.cpp [hunk 2]
```diff
@@ -39,520 +39,625 @@
...
+//
+//if(abortion_counter_ > max_abortions_)
+//{
+//    abortion_counter_ = max_abortions_;
+//    this->publishHoldTwist(period);
+//}
+//}
+//
+//ht_.hold = abortion_counter_ >= max_abortions_;
+//publishTwist(period, !ht_.hold); // if not publishing then just update data!
```
Comment (fmessmer) [inline cob_frame_tracker/src/cob_frame_tracker.cpp]: What function does this code block belong to? Remove?

File: cob_obstacle_distance/CMakeLists.txt [hunk 3]
```diff
@@ -23,25 +23,20 @@ add_message_files(
...
-)
-
 generate_messages(
   DEPENDENCIES
   geometry_msgs
 )
 
 catkin_package(
-  CATKIN_DEPENDS cob_srvs dynamic_reconfigure eigen_conversions geometry_msgs kdl_conversions kdl_parser message_runtime moveit_msgs roscpp roslib sensor_msgs shape_msgs std_msgs tf_conversions tf urdf visualization_msgs
+  CATKIN_DEPENDS cob_srvs dynamic_reconfigure eigen_conversions geometry_msgs interactive_markers kdl_conversions kdl_parser message_runtime moveit_msgs roscpp roslib sensor_msgs shape_msgs std_msgs tf_conversions tf urdf visualization_msgs
```
Comment (fmessmer) [inline cob_obstacle_distance/CMakeLists.txt]: `interactive_markers` should only be `exec_depend`

File: cob_obstacle_distance/package.xml [hunk 4]
```diff
@@ -22,6 +22,7 @@
   <depend>eigen</depend>
   <depend>fcl</depend>
   <depend>geometry_msgs</depend>
+  <depend>interactive_markers</depend>
```
Comment (fmessmer) [inline cob_obstacle_distance/package.xml]: make `exec_depend`

File: cob_obstacle_distance/scripts/test_obstacle_publisher_node.py [hunk 5]
```diff
@@ -79,19 +79,22 @@
     y = CollisionObject()
     y.id = "Funny Mesh"
     y.header.frame_id = root_frame
-    y.type.db = "package://cob_gazebo_objects/Media/models/milk.dae"
-    y.operation = CollisionObject.ADD
+     # y.type.db = "package://cob_gazebo_objects/Media/models/milk.dae"
+    y.type.db = "package://cob_twist_controller/files/torus_0_25_inner_rad.stl"
+    # y.type.db = "package://cob_twist_controller/files/untitled.dae"
+    # y.operation = CollisionObject.ADD
+    y.operation = CollisionObject.REMOVE
```
Comment (fmessmer) [inline cob_obstacle_distance/scripts/test_obstacle_publisher_node.py]: Please undo the changes in this file as it is meant to be a basic example.
Add a new one for adding the torus! If you do so, add the mesh file, too!

File: cob_obstacle_distance/src/distance_manager.cpp [hunk 6]
```diff
@@ -438,6 +425,12 @@ void DistanceManager::registerObstacle(const moveit_msgs::CollisionObject::Const
         return;
     }
 
+    if(msg->operation == msg->ADD && this->obstacle_mgr_->count(msg->id) > 0)
+    {
+        ROS_ERROR_STREAM("Element " << msg->id << " exists already. ADD not allowed!");
```
Comment (fmessmer) [inline cob_obstacle_distance/src/distance_manager.cpp:428]: According to http://docs.ros.org/jade/api/moveit_msgs/html/msg/CollisionObject.html

> Adds the object to the planning scene. If the object previously existed, it is replaced.

Thus this should be handled as "move"?

File: cob_obstacle_distance/src/frame_to_collision.cpp [hunk 7]
```diff
@@ -111,26 +111,15 @@ bool FrameToCollision::initSelfCollision(XmlRpc::XmlRpcValue& self_collision_par
...
-                    it->second.push_back(c_link->name);
-                }
-            }
+            ROS_INFO_STREAM("Create self-collision obstacle for: " << it->first);
 
             // Create real obstacles now.
             PtrIMarkerShape_t ptr_obstacle;
             this->getMarkerShapeFromUrdf(Eigen::Vector3d(), Eigen::Quaterniond(), it->first, ptr_obstacle);
+            //ptr_obstacle->setDrawable(false); // do not draw the marker in rviz again (are available in robot model->collision enabled...
+            ptr_obstacle->setDrawable(true); // do not draw the marker in rviz again (are available in robot model->collision enabled...
```
Comment (fmessmer) [inline cob_obstacle_distance/src/frame_to_collision.cpp]: So which of the two lines can be removed?

File: cob_obstacle_distance/src/marker_shapes/marker_shapes_impl.cpp [hunk 8]
```diff
@@ -48,11 +48,12 @@ MarkerShape<BVH_RSS_t>::MarkerShape(const std::string& root_frame,
         fcl::Vec3f v2(mesh.vertices[v_idx_2].x, mesh.vertices[v_idx_2].y, mesh.vertices[v_idx_2].z);
         fcl::Vec3f v3(mesh.vertices[v_idx_3].x, mesh.vertices[v_idx_3].y, mesh.vertices[v_idx_3].z);
 
-        this->fcl_bvh_.addTriangle(v1, v2, v3);
+        //this->fcl_bvh_.addTriangle(v1, v2, v3);
```
Comment (fmessmer) [inline cob_obstacle_distance/src/marker_shapes/marker_shapes_impl.cpp]: remove commented line

File: cob_obstacle_distance/src/debug/debug_obstacle_distance_node.cpp [hunk 9]
```diff
@@ -98,23 +118,30 @@ std::string chain_base_link_;
...
             marker_distance.type = visualization_msgs::Marker::TEXT_VIEW_FACING;
-            marker_distance.lifetime = ros::Duration();
+            marker_distance.lifetime = ros::Duration(0.5);
             marker_distance.action = visualization_msgs::Marker::ADD;
-            marker_distance.ns = msg->distances[i].header.frame_id;
+            marker_distance.ns = it->first;
             marker_distance.id = 69;
-            marker_distance.header.frame_id = msg->distances[i].header.frame_id;
-            marker_distance.text = boost::lexical_cast<std::string>(boost::format("%.2f") % msg->distances[i].distance);
+            // marker_distance.header.frame_id = it->second.frame_of_interest;
```
Comment (fmessmer) [inline cob_obstacle_distance/src/debug/debug_obstacle_distance_node.cpp]: remove commented line?

File: cob_obstacle_distance/src/distance_manager.cpp [hunk 10]
```diff
@@ -372,11 +354,16 @@ void DistanceManager::transformSelfCollisionFrames(const std::string frame_name)
                 tf::StampedTransform stamped_transform;
                 geometry_msgs::Transform msg_transform;
                 tf_listener_.lookupTransform(root_frame_id_, frame_name, time, stamped_transform);
-                tf::transformTFToMsg(stamped_transform, msg_transform);
+                // tf::transformTFToMsg(stamped_transform, msg_transform);
```
Comment (fmessmer) [inline cob_obstacle_distance/src/distance_manager.cpp]: remove commented line?

File: cob_twist_controller/cfg/TwistController.cfg [hunk 11]
```diff
@@ -91,7 +91,7 @@ limits = gen.add_group("Limits", "limits")
 limits.add("keep_direction",     bool_t,   0, "With keep_direction the whole joint positions and velocities vector is affected by a scaling factor. Else only individual components of the vectors are affected -> direction will be changed.", True)
 limits.add("enforce_vel_limits", bool_t,   0, "If 'True', enforce joint velocity limits in TwistController, else let RobotHW handle the limits",  True)
 limits.add("enforce_pos_limits", bool_t,   0, "If 'True', enforce joint position limits in TwistController, else let RobotHW handle the limits",  True)
-limits.add("tolerance",          double_t, 0, "Tolerance around joint position limits where velocities are normalized [degrees]",  5, 0, 15)
+limits.add("tolerance",          double_t, 0, "Tolerance around joint position limits where velocities are normalized [degrees]. Must not be zero else DIV/0!",  5, 0.001, 15)
```
Comment (fmessmer) [inline cob_twist_controller/cfg/TwistController.cfg]: Can we handle it in a way that 0.0 means `enforce_X_limits` = False? Directly in the code?

File: cob_twist_controller/include/cob_twist_controller/constraints/constraint_impl.h [hunk 12]
```diff
@@ -99,13 +105,15 @@ std::set<ConstraintBase_t> ConstraintsBuilder<PRIO>::createConstraints(const Twi
     if(CA_ON == twist_controller_params.constraint_ca)
     {
         typedef CollisionAvoidance<ConstraintParamsCA, PRIO> CollisionAvoidance_t;
-        uint32_t available_dists = data_mediator.obstacleDistancesCnt();
+        //uint32_t available_dists = data_mediator.obstacleDistancesCnt();
```
Comment (fmessmer) [inline cob_twist_controller/include/cob_twist_controller/constraints/constraint_impl.h]: remove commented line?

File: cob_twist_controller/package.xml [hunk 13]
```diff
@@ -1,7 +1,14 @@
...
+  <description>
+  The main purpose of the cob_twist_controller is to convert target twists into joint velocities. 
+  Therefore it makes use of several implemented inverse kinematics approaches at the first order differential level. 
+  The inverse differential kinematics solver considers kinematic chain extensions, singularity robustness, 
+  redundancy resolution and priority-based methods.
+  To avoid hardware destruction there is a limiter interface active as well. 
+  Via parameter server users can dynamically configure the solving strategy. 
+  </description>
 
   <author email="fxm@ipa.fhg.de">Felix Messmer</author>
```
Comment (fmessmer) [inline cob_twist_controller/package.xml:13]: @ipa-fxm-mb @ipa-fxm-cm and @ipa-fxm-fm should be added here

File: cob_twist_controller/src/constraint_solvers/solvers/stack_of_tasks_solver.cpp [hunk 14]
```diff
@@ -19,116 +19,158 @@
...
+
+    if(cstate.isTransition())
+    {
+        if(cstate.getCurrent() == CRITICAL)
+        {
+            // Eigen::MatrixXd J_task0 = q_dot_0.transpose();
+            // "global" weighting k_H for all constraint tasks.
+            //Eigen::VectorXd task = this->params_.k_H * activation_gain * magnitude * derivative_value * Eigen::VectorXd::Identity(1, 1);
+            Task_t t = (*it)->createTask();
+            //double factor = this->params_.k_H * activation_gain * std::abs(magnitude);
```
Comment (fmessmer) [inline cob_twist_controller/src/constraint_solvers/solvers/stack_of_tasks_solver.cpp]: remove commented lines above?

File: cob_cartesian_controller/scripts/action_test.py (hunk 1 above)
Comment (ipa-fxm-mb) [inline cob_cartesian_controller/scripts/action_test.py]: Removed!

File: cob_frame_tracker/src/cob_frame_tracker.cpp (hunk 2 above)
Comment (ipa-fxm-mb) [inline cob_frame_tracker/src/cob_frame_tracker.cpp]: Removed!

File: cob_obstacle_distance/scripts/test_obstacle_publisher_node.py (hunk 5 above)
Comment (ipa-fxm-mb) [inline cob_obstacle_distance/scripts/test_obstacle_publisher_node.py]: There is an example file already which is unchanged. 

Added stl file.

File: cob_obstacle_distance/src/frame_to_collision.cpp (hunk 7 above)
Comment (ipa-fxm-mb) [inline cob_obstacle_distance/src/frame_to_collision.cpp]: Fixed!

File: cob_obstacle_distance/src/marker_shapes/marker_shapes_impl.cpp (hunk 8 above)
Comment (ipa-fxm-mb) [inline cob_obstacle_distance/src/marker_shapes/marker_shapes_impl.cpp]: Fixed!

File: cob_obstacle_distance/src/debug/debug_obstacle_distance_node.cpp (hunk 9 above)
Comment (ipa-fxm-mb) [inline cob_obstacle_distance/src/debug/debug_obstacle_distance_node.cpp]: Removed!

File: cob_obstacle_distance/src/distance_manager.cpp (hunk 10 above)
Comment (ipa-fxm-mb) [inline cob_obstacle_distance/src/distance_manager.cpp]: Removed!

File: cob_twist_controller/include/cob_twist_controller/constraints/constraint_impl.h (hunk 12 above)
Comment (ipa-fxm-mb) [inline cob_twist_controller/include/cob_twist_controller/constraints/constraint_impl.h]: Removed!

File: cob_twist_controller/package.xml (hunk 13 above)
Comment (ipa-fxm-mb) [inline cob_twist_controller/package.xml:13]: Added to authors.

File: cob_twist_controller/src/constraint_solvers/solvers/stack_of_tasks_solver.cpp (hunk 14 above)
Comment (ipa-fxm-mb) [inline cob_twist_controller/src/constraint_solvers/solvers/stack_of_tasks_solver.cpp]: Removed!

File: cob_twist_controller/cfg/TwistController.cfg (hunk 11 above)
Comment (ipa-fxm-mb) [inline cob_twist_controller/cfg/TwistController.cfg]: Added: 

```
if(twist_controller_params_.tolerance <= DIV0_SAFE)
{
    ROS_ERROR("The tolerance for enforce limits is smaller than DIV/0 threshold. Therefore both enforce_limits are set to false!");
    twist_controller_params_.enforce_pos_limits = config.enforce_pos_limits = false;
    twist_controller_params_.enforce_vel_limits = config.enforce_vel_limits = false;
}
```

**Comment by fmessmer** [comment]
All things considered (or documented in follow up issues #52 #53)
Good to merge!
